### PR TITLE
fix: allow snapcraft `apps: directive` to use custom commands

### PR DESCRIPTION
**Pull request by jaredfolkins**
enhc: allow snapcraft apps: directive to utilize custom commands, defaults to binary name if no apps: are specified

This patch fixes the ambiguity when packaging snaps. By removing the ambiguity it explicitly allows custom apps: directives. This *fixes* the yet *unconfirmed* side-effect/bug where command: flags could go missing.

Slight refactoring of tests for specific conditions no longer apply.

Feedback/Further Refactoring is welcome.

Snapcraft: Lost flags or naming ambiguity #1263

### CODE REVIEW

**Comment by jaredfolkins** [comment]
The change to the tests is because the logic changed a bit.

pseudo code
```
If NO apps: directive is specified 

  ONLY then create a default apps: directive using the binary name.

else

  we utilize what has been specified in the apps: directive

```

Example

```
##
## BUILDS
## 
builds:
##
## kushtakad
##
- id: kushtakad
  main: ./main.go
  binary: kushtakad
  goos:
    - darwin
    - linux
    - windows
  goarch:
    - amd64
  env:
    - GO111MODULE=on
    - CGO_ENABLED=0

##
## BEFORE
## 
before:
  hooks:
    - go mod download
    - packr2 clean
    - packr2
      
checksum:
  name_template: 'checksums.txt'

snapshot:
  name_template: "{{ .Tag }}-beta"

changelog:
  sort: asc

##
## ARCHIVES
##
archives:
  ##
  ## kushtakad
  ##
  - id: kushtakad
    builds:
    - kushtakad
    format: gz
    format_overrides:
      - goos: windows
        format: zip
    files:
    - none*

##
## SNAPS
##
snapcrafts:
  - id: kushtakad

    name: kushtakad
    summary: Kushtakad
    description: Kushtaka.org is a sustainable all-in-one honeypot and honeytoken orchestrator for under-resourced Blue Teams. 
    license: Apache-2.0
    publish: false
    grade: beta
    confinement: strict
    name_template: "{{ .ProjectName }}_{{ .Version }}_{{ .Os }}_{{ .Arch }}"
    replacements:
      amd64: 64-bit
      darwin: macOS
      linux: Tux

    apps:
      set:
        args: -set
      sensor:
        plugs: ["network", "network-bind"]
        daemon: simple
        args: -sensor
      server:
        plugs: ["network", "network-bind"]
        daemon: simple

    #grade: edge
    #base: core18
    #confinement: classic

```

**Comment by jaredfolkins** [comment]
This allows us to also follow the snapcraft naming conventions under `snap services` such as `kushtakad.sensor` or `kushtakad.server` or `binaryname.daemon`.

**Comment by jaredfolkins** [comment]
```
$ sudo snap install --devmode dist/kushtakad_v0.2.2-beta_Tux_64-bit.snap
kushtakad v0.2.2-beta installed
$ sudo snap services
Service           Startup  Current   Notes
kushtakad.sensor  enabled  inactive  -
kushtakad.server  enabled  active    -
$
```

**Comment by jaredfolkins** [comment]
It appears the reason that the travisci tests are failing is because of the old logic. Goreleaser was in many cases still adding a default apps: directive even when one was specified. This will clutter up user's cli and UX especially when you don't want your users executing the base binary.

Also on my Bug report, you can see that I had to do all sorts of weird hacky stuff just to get snapcraft to work. But it definitely wasn't the correct way to do it.

**Comment by caarlos0** [comment]
I think we can then fix the tests as well?

**Comment by caarlos0** [comment]
looks good overall, just need to make the build pass :) 

thanks for the PR 🚀

**Comment by jaredfolkins** [comment]
Hey @caarlos0 ,

On my ubuntu 18.04 machine, these tests pass. On TravisCI, they fail. Ideas?

```
2019/12/17 16:11:14  info creating                  arch=i386 snap=/tmp/archivetest142048546/dist/foo_386.snap
2019/12/17 16:11:14  info creating                  arch=armhf snap=/tmp/archivetest142048546/dist/foo_arm6.snap
2019/12/17 16:11:14  info creating                  arch=amd64 snap=/tmp/archivetest142048546/dist/foo_amd64.snap
2019/12/17 16:11:14  info creating                  arch=armhf snap=/tmp/archivetest142048546/dist/foo_and_bar_arm6.snap
2019/12/17 16:11:14  info creating                  arch=i386 snap=/tmp/archivetest142048546/dist/foo_and_bar_386.snap
2019/12/17 16:11:14  info creating                  arch=amd64 snap=/tmp/archivetest142048546/dist/foo_and_bar_amd64.snap
2019/12/17 16:11:15  info creating                  arch=armhf snap=/tmp/archivetest142048546/dist/bar_arm6.snap
2019/12/17 16:11:15  info creating                  arch=amd64 snap=/tmp/archivetest142048546/dist/bar_amd64.snap
2019/12/17 16:11:15  info creating                  arch=i386 snap=/tmp/archivetest142048546/dist/bar_386.snap
2019/12/17 16:11:15  info creating                  arch=armhf snap=/tmp/archivetest717655204/dist/foo_arm6.snap
2019/12/17 16:11:15  info creating                  arch=amd64 snap=/tmp/archivetest717655204/dist/foo_amd64.snap
2019/12/17 16:11:15  info creating                  arch=i386 snap=/tmp/archivetest717655204/dist/foo_386.snap
2019/12/17 16:11:16  info creating                  arch=armhf snap=/tmp/archivetest917773747/dist/foo_arm6.snap
2019/12/17 16:11:16  info creating                  arch=i386 snap=/tmp/archivetest917773747/dist/foo_386.snap
2019/12/17 16:11:16  info creating                  arch=amd64 snap=/tmp/archivetest917773747/dist/foo_amd64.snap
2019/12/17 16:11:16  info creating                  arch=armhf snap=/tmp/archivetest291485558/dist/foo_arm6.snap
2019/12/17 16:11:16  info creating                  arch=i386 snap=/tmp/archivetest291485558/dist/foo_386.snap
2019/12/17 16:11:16  info creating                  arch=amd64 snap=/tmp/archivetest291485558/dist/foo_amd64.snap
2019/12/17 16:11:17  info creating                  arch=armhf snap=/tmp/archivetest110808157/dist/foo_arm6.snap
2019/12/17 16:11:17  info creating                  arch=amd64 snap=/tmp/archivetest110808157/dist/foo_amd64.snap
2019/12/17 16:11:17  info creating                  arch=i386 snap=/tmp/archivetest110808157/dist/foo_386.snap
2019/12/17 16:11:17  info creating                  arch=armhf snap=/tmp/archivetest885875224/dist/foo_arm6.snap
2019/12/17 16:11:17  info creating                  arch=amd64 snap=/tmp/archivetest885875224/dist/foo_amd64.snap
2019/12/17 16:11:17  info creating                  arch=i386 snap=/tmp/archivetest885875224/dist/foo_386.snap
2019/12/17 16:11:18  info pushing snap              snap=mybin
2019/12/17 16:11:18  warn DEPRECATED: `snapcraft` should not be used anymore, check https://goreleaser.com/deprecations#snapcraft for more info.
PASS
ok  	github.com/goreleaser/goreleaser/internal/pipe/snapcraft	4.262s
```

**Comment by caarlos0** [comment]
will try your patch locally to see what I can find

**Comment by caarlos0** [comment]
it works locally as well... I've seen that locally I have snapcraft 3.9 and on travis its using 3.8, can you check which version you have locally as well?

**Comment by jaredfolkins** [comment]
```
snap    2.42.5
snapd   2.42.5
series  16
ubuntu  18.04
kernel  4.15.0-72-generic
```

**Comment by jaredfolkins** [comment]
Yeah, I have two boxes with newer snap 3.9 and 2.42.5 and it works.

**Comment by caarlos0** [comment]
Not stale

**Comment by caarlos0** [comment]
I changed how snapcraft is installed and now this works.

Issue now is just linting, will merge and fix on master!

thanks @jaredfolkins

File: internal/pipe/snapcraft/snapcraft.go [hunk 1]
```diff
@@ -185,34 +186,26 @@ func create(ctx *context.Context, snap config.Snapcraft, arch string, binaries [
...
-			}
+		metadata.Apps[snap.Name] = AppMetadata{
+			Command: filepath.Base(binaries[0].Name),
 		}
-		metadata.Apps[name] = appMetadata
-		metadata.Plugs = snap.Plugs
+	}
 
+	for _, binary := range binaries {
+
```
Comment (caarlos0) [inline internal/pipe/snapcraft/snapcraft.go]:
```suggestion
```

File: internal/pipe/snapcraft/snapcraft.go [hunk 2]
```diff
@@ -232,11 +225,36 @@ func create(ctx *context.Context, snap config.Snapcraft, arch string, binaries [
...
+				if configAppMetadata.Completer != "" {
+					completerPath = configAppMetadata.Completer
+					appMetadata.Completer = filepath.Base(completerPath)
+				}
+			}
+
+			metadata.Apps[name] = appMetadata
+			metadata.Plugs = snap.Plugs
+
```
Comment (caarlos0) [inline internal/pipe/snapcraft/snapcraft.go]:
```suggestion
```

**Pull request merged**